README.md
# Mojolicious::Plugin::Ident [![Build Status](https://secure.travis-ci.org/plicease/Mojolicious-Plugin-Ident.png)](http://travis-ci.org/plicease/Mojolicious-Plugin-Ident)

Mojolicious plugin to interact with a remote ident service

# SYNOPSIS

    use Mojolicious::Lite;
    plugin 'ident';
    
    # log the ident user for every connection (async ident)
    under sub {
      shift->ident(sub {
        my $id_res = shift; # $id_res isa Mojolicious::Plugin::Ident::Response
        if($id_res->is_success) {
          app->log->info("ident user is " . $id_res->username);
        } else {
          app->log->info("unable to ident remote user");
        }
      });

      1;
    };
    
    # get the username of the remote using ident protocol
    get '/' => sub {
      my $self = shift;
      my $id_res = $self->ident; # $id_res isa Mojolicious::Plugin::Ident::Response
      $self->render(text => "hello " . $id_res->username);
    };
    
    # only allow access to the user on localhost which
    # started the mojolicious lite app with non-blocking
    # ident call (requires Mojolicious 4.28)
    under sub {
      my($self) = @_;
      $self->ident_same_user(sub {
        my($same) = @_;
        unless($same) {
          return $self->render(
            text   => 'permission denied',
            status => 403,
          );
        }
        $self->continue;
      });
      return undef;
    };
    
    get '/private' => sub {
      shift->render(text => "secret place");
    };
    
    # only allow access to the user on localhost which 
    # started the mojolicious lite app (all versions of
    # Mojolicious)
    under sub {
      my($self) = @_;
      if($self->ident_same_user) {
        return 1;
      } else {
        $self->render(
          text   => 'permission denied',
          status => 403,
        );
      }
    };
    
    get '/private' => sub {
      shift->render(text => "secret place");
    };

# DESCRIPTION

This plugin provides an interface for querying an ident service on a 
remote system.  The ident protocol helps identify the user of a 
particular TCP connection.  If the remote client connecting to your 
Mojolicious application is running the ident service you can identify 
the remote users' name.  This can be useful for determining the source 
of abusive or malicious behavior.  Although ident can be used to 
authenticate users, it is not recommended for untrusted networks and 
systems (see CAVEATS below).

Under the covers this plugin uses [AnyEvent::Ident](https://metacpan.org/pod/AnyEvent::Ident).

# OPTIONS

## timeout

    plugin 'ident' => { timeout => 60 };

Default number of seconds to wait before timing out when contacting the remote
ident server.  The default is 2.

## port

    plugin 'ident' => { port => 113 };

Port number to connect to.  Usually this will be 113, but you may want to change
this for testing or some other purpose.

# HELPERS

## ident \[ $tx, \[ $timeout \] \], \[ $callback \]

This helper makes a ident request.  This helper takes two optional arguments,
a transaction `$tx` and a timeout `$timeout`.  If not specified, the current
transaction and the configured default timeout will be used.  If a callback
is provided then the request is non-blocking.  If no callback is provided,
it will block until a response comes back or the timeout expires.

With a callback (non-blocking):

    get '/' => sub {
      my $self = shift;
      $self->ident(sub {
        my $res = shift->res;
        if($res->is_success)
        {
          $self->render(text =>
            "username: " . $res->username .
            "os:       " . $res->os
          );
        }
        else
        {
          $self->render(text =>
            "error: " . $res->error_type
          );
        }
      };
    };

The callback is passed an instance of [Mojolicious::Plugin::Ident::Response](https://metacpan.org/pod/Mojolicious::Plugin::Ident::Response).  Even if
the response is an error.  The `is_success` method on [Mojolicious::Plugin::Ident::Response](https://metacpan.org/pod/Mojolicious::Plugin::Ident::Response)
will tell you if the response is an error or not.

Without a callback (blocking):

    get '/' => sub {
      my $self = shift;
      my $ident = $self->ident;
      $self->render(text =>
        "username: " . $ident->username .
        "os:       " . $ident->os
      );
    };

Returns an instance of [Mojolicious::Plugin::Ident::Response](https://metacpan.org/pod/Mojolicious::Plugin::Ident::Response), which 
provides two fields, username and os for the remote connection.

When called in blocking mode (without a callback), the ident helper will throw 
an exception if

- it cannot connect to the remote's ident server
- the connection to the remote's ident server times out
- the remote ident server returns an error

    under sub { eval { shift->ident->same_user } };
    get '/private' => 'private_route';

The ident response class also has a same\_user method which can be used
to determine if the user which started the Mojolicious application and
the remote user are the same.  The user is considered the same if the
remote connection came over the loopback address (127.0.0.1) and the
username matches either the server's username or real UID.  Although
this can be used as a simple authentication method, keep in mind that it
may not be secure (see CAVEATS below).

## ident\_same\_user \[ $tx, \[ $timeout \] \], \[ $callback \]

This helper makes an ident request and attempts to determine if the 
user that made the request is the same as the one that started the
Mojolicious application.  This helper takes two optional arguments,
a transaction `$tx` and a timeout `$timeout`.  If not specified, the current
transaction and the configured default timeout will be used.  If a callback
is provided then the request is non-blocking.  If no callback is provided,
it will block until a response comes back or the timeout expires.

With a callback (non-blocking):

    get '/private' => sub {
      my $self = shift;
      $self->ident_same_user(sub {
        my $same_user = shift;
        $same_user ? $self->render(text => 'private text') : $self->reply->not_found;
      });
    }

When the response comes back it will call the callback and pass in a boolean
value indicating if the user is the same.  If the ident request connects
and does not timeout, then result will be cached.  If cached the callback may
be called immediately, before re-entering the event loop.

Without a callback (blocking):

    under sub { shift->ident_same_user };
    get '/private' => 'private_route';

without a callback this helper will return true or false depending on
if the user is the same.  It should never throw an exception.

# CAVEATS

[The RFC for the ident protocol](http://tools.ietf.org/html/rfc1413)
clearly states that ident should not be used for authentication, at
most it should be used only for audit (for example annotating log
files).

In Windows and possibly other operating systems, an unprivileged user
can listen to port 113 and on any untrusted network, a remote ident
server is not a secure authentication mechanism.  Most modern operating
systems do not enable the ident service by default, so unless you have
control both the client and the server and can configure the ident
service securely on both, its usefulness is reduced.

Using this module in the non-blocking mode requires that [AnyEvent](https://metacpan.org/pod/AnyEvent) use 
its [EV](https://metacpan.org/pod/EV) implementation, which is also used by [Mojolicious](https://metacpan.org/pod/Mojolicious), if it is 
loaded.  This shouldn't be a problem, as [EV](https://metacpan.org/pod/EV) is a prerequisite to this 
module (though it does not use it directly), and both [AnyEvent](https://metacpan.org/pod/AnyEvent) and 
[Mojolicious](https://metacpan.org/pod/Mojolicious) will prefer to use [EV](https://metacpan.org/pod/EV) if it is installed.  You do have 
to make sure that you do not force another event loop, such as 
[AnyEvent::Loop](https://metacpan.org/pod/AnyEvent::Loop), unless you are using only the blocking mode.

[Mojolicious](https://metacpan.org/pod/Mojolicious) 4.28 introduced support for non-blocking operations in bridges.
Prior to that if a bridge returned false the server would generate a
404 "Not Found" reply.  In 4.29 a bridge returning false would not render
anything and thus timeout if the bridge didn't render anything.  Thus in
older versions of [Mojolicious](https://metacpan.org/pod/Mojolicious) this:

    under sub { shift->ident_same_user };

would return 404 if the remote and local users are not the same.  To get the
same behavior in both new and old versions of [Mojolicious](https://metacpan.org/pod/Mojolicious):

    under sub {
      my($self) = @_;
      if($self->ident_same_user) {
        return 0;
      } else {
        $self->reply->not_found;
        return 1;
      }
    };

Most of the time you should really return a 403, instead of not found (as in
the synopsis above), but this is what you would want to do if you wanted a
resource to be invisible and unavailable rather than just unavailable to the
wrong user.

I only mention this because old versions of this plugin had documentation
which included the older form in its synopsis.

# AUTHOR

Graham Ollis <plicease@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Mojolicious::Plugin::Ident

{{$NEXT}}

0.31      2015-02-28 10:42:10 -0500
  - Mojo 6.x compat

0.30      2013-09-02 09:52:17 -0400
  - Mojolicious 4.28 compatability

0.29      2013-05-17 12:04:57 America/New_York
  - Mojolicious 4.0 compatability

0.28      2013-02-09 18:53:06 America/New_York
  - rm dep base

0.27      2013-02-09 09:55:14 America/New_York
  - compat with AnyEvent::Ident 0.04

0.26      2013-01-19 09:00:51 America/New_York
  - Make EV (not AE) a prereq (non-blocking mode does
    not work with other event loops)
  - documentation updates

0.25      2013-01-17 13:06:44 America/New_York
  - Make AE a prereq (non-blocking mode does
    not work with AnyEvent::Loop)

0.24      2013-01-17 12:43:09 America/New_York
  - Add additional diagnostics to non_blocking test

0.23      2013-01-17 12:00:39 America/New_York
  - Make M::P::I::Response inherit from AE::I::Response
    in order to simplify the interface and make non-blocking
    more like blocking mode

0.22      2013-01-15 11:49:31 America/New_York
  - Migrate from Net::Ident to AnyEvent::Ident

0.21      2012-11-29 12:15:36 America/New_York
  - remove extra deps

0.2       2012-10-02 15:32:10 America/New_York
  - fix $VERSION so that it is not undef

0.1       2012-09-07 12:14:02 America/New_York
  - initial version


example/ident.pl
#!/usr/bin/env perl

use Mojolicious::Lite;

plugin 'ident';

# helper to get the controller
helper c => sub { shift };

get '/' => 'index';

under sub {
  my($self) = @_;

  return 1 if $self->ident_same_user;

  $self->render(status => 403, template => 403);
  return;
};

get '/private' => 'private';

app->start;
__DATA__

@@ index.html.ep
% layout 'default';
% title 'ident test';
<table>
  <tr>
    <td>username:</td>
    <td><%= ident->username %></td>
  </tr>
  </tr>
    <td>os</td>
    <td><%= ident->os %></td>
  </tr>
  <tr>
    <td>local</td>
    <td><%= c->tx->local_address %>:<%= c->tx->local_port %></td>
  </tr>
  <tr>
    <td>remote</td>
    <td><%= c->tx->remote_address %>:<%= c->tx->remote_port %></td>
  </tr>
</table>

@@ private.html.ep
% layout 'default';
% title 'ident test';
<p>you are in the private area</p>

@@ 403.html.ep
% layout 'default';
% title 'Forbidden';
<h1>403 Forbidden</h1>
<p>You do not have permission to access this resource</p>

@@ layouts/default.html.ep
<!DOCTYPE html>
<html>
  <head><title><%= title %></title></head>
  <body><%= content %></body>
</html>


lib/Mojolicious/Plugin/Ident/Response.pm
package Mojolicious::Plugin::Ident::Response;

use strict;
use warnings;
use AnyEvent::Ident::Response;
our @ISA = qw( AnyEvent::Ident::Response );

# ABSTRACT: Ident response object
# VERSION

=head1 DESCRIPTION

This class represents the responses as they come back
from the remote ident server.

L<Mojolicious::Plugin::Ident::Response> is a L<AnyEvent::Ident::Response>.
The base class is subject to change in a future version, although it
this class will always provide the interface documented here.

=head1 ATTRIBUTES

=head2 $ident-E<gt>username

The username of the remote connection as provided by
the remote ident server.

=head2 $ident-E<gt>os

The operating system of the remote connection as provided
by the remote ident server.

=head2 $ident-E<gt>is_success

True if the ident response was not an error.  Only
useful in non-blocking mode, as in blocking mode an
exception will be thrown in the case of error.

=head2 $ident-E<gt>error_type

The error type returned by the ident server, if an error
happened.  Only useful in non-blocking mode, as in blocking
mode an exception will be thrown in the case of error.

=cut

my $server_user_uid;
my $server_user_name;

sub _server_user_uid  { $server_user_uid  }
sub _server_user_name { $server_user_name }

sub _setup
{
  if($^O eq 'MSWin32')
  {
    $server_user_name = $ENV{USERNAME};
  }
  else
  {
    $server_user_uid  = $<;
    $server_user_name = scalar getpwuid($<);
  }
  die "could not determine username"
    unless defined $server_user_name
    &&     $server_user_name;
}

=head1 METHODS

=head2 $ident-E<gt>same_user

Returns true if the remote user is the same as the one which started the 
Mojolicious application.  The user is considered the same if the remote 
connection came over the loopback address (127.0.0.1) and the username 
matches either the server's username or real uid.

=cut

sub same_user
{
  my($self) = @_;
  return unless $self->{remote_address} eq '127.0.0.1';
  return 1 if $self->username eq $server_user_name;
  return 1 if defined $server_user_uid && $self->username =~ /^\d+$/ && $self->username == $server_user_uid;
  return;
}

1;

=head1 SEE ALSO

L<Mojolicious::Plugin::Ident>,
L<AnyEvent::Ident::Response>

=cut


lib/Mojolicious/Plugin/Ident.pm
package Mojolicious::Plugin::Ident;

use strict;
use warnings;
use v5.10;
use Mojo::Base 'Mojolicious::Plugin';
use AnyEvent;
use AnyEvent::Ident::Client;
use Mojo::Exception;
use Mojolicious::Plugin::Ident::Response;

# ABSTRACT: Mojolicious plugin to interact with a remote ident service
# VERSION

=head1 SYNOPSIS

 use Mojolicious::Lite;
 plugin 'ident';
 
 # log the ident user for every connection (async ident)
 under sub {
   shift->ident(sub {
     my $id_res = shift; # $id_res isa Mojolicious::Plugin::Ident::Response
     if($id_res->is_success) {
       app->log->info("ident user is " . $id_res->username);
     } else {
       app->log->info("unable to ident remote user");
     }
   });

   1;
 };
 
 # get the username of the remote using ident protocol
 get '/' => sub {
   my $self = shift;
   my $id_res = $self->ident; # $id_res isa Mojolicious::Plugin::Ident::Response
   $self->render(text => "hello " . $id_res->username);
 };
 
 # only allow access to the user on localhost which
 # started the mojolicious lite app with non-blocking
 # ident call (requires Mojolicious 4.28)
 under sub {
   my($self) = @_;
   $self->ident_same_user(sub {
     my($same) = @_;
     unless($same) {
       return $self->render(
         text   => 'permission denied',
         status => 403,
       );
     }
     $self->continue;
   });
   return undef;
 };
 
 get '/private' => sub {
   shift->render(text => "secret place");
 };
 
 # only allow access to the user on localhost which 
 # started the mojolicious lite app (all versions of
 # Mojolicious)
 under sub {
   my($self) = @_;
   if($self->ident_same_user) {
     return 1;
   } else {
     $self->render(
       text   => 'permission denied',
       status => 403,
     );
   }
 };
 
 get '/private' => sub {
   shift->render(text => "secret place");
 };

=head1 DESCRIPTION

This plugin provides an interface for querying an ident service on a 
remote system.  The ident protocol helps identify the user of a 
particular TCP connection.  If the remote client connecting to your 
Mojolicious application is running the ident service you can identify 
the remote users' name.  This can be useful for determining the source 
of abusive or malicious behavior.  Although ident can be used to 
authenticate users, it is not recommended for untrusted networks and 
systems (see CAVEATS below).

Under the covers this plugin uses L<AnyEvent::Ident>.

=head1 OPTIONS

=head2 timeout

 plugin 'ident' => { timeout => 60 };

Default number of seconds to wait before timing out when contacting the remote
ident server.  The default is 2.

=head2 port

 plugin 'ident' => { port => 113 };

Port number to connect to.  Usually this will be 113, but you may want to change
this for testing or some other purpose.

=head1 HELPERS

=head2 ident [ $tx, [ $timeout ] ], [ $callback ]

This helper makes a ident request.  This helper takes two optional arguments,
a transaction C<$tx> and a timeout C<$timeout>.  If not specified, the current
transaction and the configured default timeout will be used.  If a callback
is provided then the request is non-blocking.  If no callback is provided,
it will block until a response comes back or the timeout expires.

With a callback (non-blocking):

 get '/' => sub {
   my $self = shift;
   $self->ident(sub {
     my $res = shift->res;
     if($res->is_success)
     {
       $self->render(text =>
         "username: " . $res->username .
         "os:       " . $res->os
       );
     }
     else
     {
       $self->render(text =>
         "error: " . $res->error_type
       );
     }
   };
 };

The callback is passed an instance of L<Mojolicious::Plugin::Ident::Response>.  Even if
the response is an error.  The C<is_success> method on L<Mojolicious::Plugin::Ident::Response>
will tell you if the response is an error or not.

Without a callback (blocking):

 get '/' => sub {
   my $self = shift;
   my $ident = $self->ident;
   $self->render(text =>
     "username: " . $ident->username .
     "os:       " . $ident->os
   );
 };

Returns an instance of L<Mojolicious::Plugin::Ident::Response>, which 
provides two fields, username and os for the remote connection.

When called in blocking mode (without a callback), the ident helper will throw 
an exception if

=over 4

=item * it cannot connect to the remote's ident server

=item * the connection to the remote's ident server times out

=item * the remote ident server returns an error

=back

 under sub { eval { shift->ident->same_user } };
 get '/private' => 'private_route';

The ident response class also has a same_user method which can be used
to determine if the user which started the Mojolicious application and
the remote user are the same.  The user is considered the same if the
remote connection came over the loopback address (127.0.0.1) and the
username matches either the server's username or real UID.  Although
this can be used as a simple authentication method, keep in mind that it
may not be secure (see CAVEATS below).

=head2 ident_same_user [ $tx, [ $timeout ] ], [ $callback ]

This helper makes an ident request and attempts to determine if the 
user that made the request is the same as the one that started the
Mojolicious application.  This helper takes two optional arguments,
a transaction C<$tx> and a timeout C<$timeout>.  If not specified, the current
transaction and the configured default timeout will be used.  If a callback
is provided then the request is non-blocking.  If no callback is provided,
it will block until a response comes back or the timeout expires.

With a callback (non-blocking):

 get '/private' => sub {
   my $self = shift;
   $self->ident_same_user(sub {
     my $same_user = shift;
     $same_user ? $self->render(text => 'private text') : $self->reply->not_found;
   });
 }

When the response comes back it will call the callback and pass in a boolean
value indicating if the user is the same.  If the ident request connects
and does not timeout, then result will be cached.  If cached the callback may
be called immediately, before re-entering the event loop.

Without a callback (blocking):

 under sub { shift->ident_same_user };
 get '/private' => 'private_route';

without a callback this helper will return true or false depending on
if the user is the same.  It should never throw an exception.

=head1 CAVEATS

L<The RFC for the ident protocol|http://tools.ietf.org/html/rfc1413>
clearly states that ident should not be used for authentication, at
most it should be used only for audit (for example annotating log
files).

In Windows and possibly other operating systems, an unprivileged user
can listen to port 113 and on any untrusted network, a remote ident
server is not a secure authentication mechanism.  Most modern operating
systems do not enable the ident service by default, so unless you have
control both the client and the server and can configure the ident
service securely on both, its usefulness is reduced.

Using this module in the non-blocking mode requires that L<AnyEvent> use 
its L<EV> implementation, which is also used by L<Mojolicious>, if it is 
loaded.  This shouldn't be a problem, as L<EV> is a prerequisite to this 
module (though it does not use it directly), and both L<AnyEvent> and 
L<Mojolicious> will prefer to use L<EV> if it is installed.  You do have 
to make sure that you do not force another event loop, such as 
L<AnyEvent::Loop>, unless you are using only the blocking mode.

L<Mojolicious> 4.28 introduced support for non-blocking operations in bridges.
Prior to that if a bridge returned false the server would generate a
404 "Not Found" reply.  In 4.29 a bridge returning false would not render
anything and thus timeout if the bridge didn't render anything.  Thus in
older versions of L<Mojolicious> this:

 under sub { shift->ident_same_user };

would return 404 if the remote and local users are not the same.  To get the
same behavior in both new and old versions of L<Mojolicious>:

 under sub {
   my($self) = @_;
   if($self->ident_same_user) {
     return 0;
   } else {
     $self->reply->not_found;
     return 1;
   }
 };

Most of the time you should really return a 403, instead of not found (as in
the synopsis above), but this is what you would want to do if you wanted a
resource to be invisible and unavailable rather than just unavailable to the
wrong user.

I only mention this because old versions of this plugin had documentation
which included the older form in its synopsis.

=cut

sub register
{
  my($self, $app, $conf) = @_;

  Mojolicious::Plugin::Ident::Response->_setup;

  my $default_timeout = $conf->{timeout} // 2;
  my $port = $conf->{port} // 113;

  $app->helper(ident => sub {
    my $callback;
    $callback = pop if ref($_[-1]) eq 'CODE';
    my($controller, $tx, $timeout) = @_;
    $tx //= $controller->tx;
    $timeout //= $default_timeout;
    
    if($callback)
    {
      AnyEvent::Ident::Client
        ->new( response_class => 'Mojolicious::Plugin::Ident::Response', hostname => $tx->remote_address, port => $port )
        ->ident($tx->remote_port, $tx->local_port, sub {
          my $ident_response = shift;
          $ident_response->{remote_address} = $tx->remote_address;
          $callback->($ident_response) 
        }
      );
      return;
    }
    
    my $done = AnyEvent->condvar;
    
    my $w = AnyEvent->timer(after => $timeout // $default_timeout, cb => sub {
      $done->croak("ident timeout");
    });
    
    my $ident_response;
    
    AnyEvent::Ident::Client
      ->new( response_class => 'Mojolicious::Plugin::Ident::Response', hostname => $tx->remote_address, port => $port )
      ->ident($tx->remote_port, $tx->local_port, sub {
        $ident_response = shift;
        $ident_response->{remote_address} = $tx->remote_address;
        $done->send(1);
      }
    );
    
    $done->recv;
    undef $w;

    if($ident_response->is_success)
    {
      return $ident_response;
    }
    else
    {
      my $error = 'ident error: ' . $ident_response->error_type;
      $controller->app->log->error($error);
      die Mojo::Exception->new($error);
    }
  });
  
  $app->helper(ident_same_user => sub {
    my $callback;
    $callback = pop if ref($_[-1]) eq 'CODE';
    my $controller = shift;
    
    if($callback)
    {
      my($tx, $timeout) = @_;
      $tx //= $controller->tx;
      $timeout //= $default_timeout;
      
      if(defined $controller->session('ident_same_user'))
      {
        Mojo::IOLoop->timer(0 => sub {
          $callback->($controller->session('ident_same_user'));
        });
      }
      else
      {
        $controller->ident($tx, $timeout, sub {
          my $res = shift;
          if($res->is_success)
          {
            my $same_user = $res->same_user;
            $controller->session('ident_same_user' => $same_user);
            $callback->($same_user);
          }
          else
          {
            $callback->(0);
          }
        });
      }
    }
    else
    {
      $controller->session('ident_same_user') // do {
        my $same_user = eval { $controller->ident(@_)->same_user };
        return if $@;
        $controller->session('ident_same_user' => $same_user);
        $same_user;
      };
    }
  });
}

1;
